### PR TITLE
BCDA-3610 - Introduce FHIR r4 models and implement v2 metadata ep

### PR DESCRIPTION
**Pull request by mtrang1263**
### Fixes [BCDA-3610](https://jira.cms.gov/browse/BCDA-3610)

To begin support for BFD v2 requests, we want to start introducing the FHIR r4 models. For this test, we'll add a v2 endpoint that will respond back with an r4 CapabilityStatement.

### Change Details

1. Introduce FHIR r4 dependency (github.com/samply/golang-fhir-models)
2. Add new v2 endpoint for metadata requests. This request responds back using the r4 datatypes.

Will plan on adding appropriate changes after https://github.com/CMSgov/bcda-app/pull/549 is merged.

### Security Implications

- [x] new software dependencies

Introducing new FHIR r4 model dependency. It can be found here: https://github.com/samply/golang-fhir-models

- [ ] security controls or supporting software altered
- [x] new data stored or transmitted

We will return FHIR r4 CapabilityStatement from this endpoint. We were using stu3 before.

- [ ] security checklist is completed for this change
- [ ] requires more information or team discussion to evaluate security implications
- [x] no PHI/PII is affected by this change

### Acceptance Validation

1. CI passes
2. Run bdt-scan to ensure the status endpoint passes as expected.

### CODE REVIEW

File: bcda/web/v2/api.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	"github.com/CMSgov/bcda-app/bcda/servicemux"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	dateFormat = "2006-01-02"
+)
+
+/*
+	swagger:route GET /api/v1/metadata metadata metadata
```
Comment (msnook) [inline bcda/web/v2/api.go:21]: v2?

File: bcda/web/v2/api.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	if servicemux.IsHTTPS(r) {
+		scheme = "https"
+	}
+	host := fmt.Sprintf("%s://%s", scheme, r.Host)
+	statement := fhir.CapabilityStatement{
+		Status:    fhir.PublicationStatusActive,
+		Date:      dt.Format(dateFormat),
+		Publisher: getStringPtr("Centers for Medicare & Medicaid Services"),
+		Kind:      fhir.CapabilityStatementKindInstance,
+		// TODO: Update to r4 once endpoint is available
```
Comment (msnook) [inline bcda/web/v2/api.go:50]: Do we need an explicit ticket to make this update, so we don't forget?

File: bcda/web/v2/api_test.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	assert.Equal(t, "application/json", res.Header.Get("Content-Type"))
+	assert.Equal(t, http.StatusOK, res.StatusCode)
+
+	resp, err := ioutil.ReadAll(res.Body)
+	assert.NoError(t, err)
+
+	cs, err := fhir.UnmarshalCapabilityStatement(resp)
+	assert.NoError(t, err)
+
+	// Expecting an R4 response so we'll evaluate some fields to reflect that
```
Comment (msnook) [inline bcda/web/v2/api_test.go:30]: Nice!

**Comment by msnook** [comment]
What are the appropriate changes associated with this comment?
```
Will plan on adding appropriate changes after #549 is merged.
```

Since this PR is ready to go in first, should we just add the scaffolding as part of #549?  Or are you envisioning a separate PR?

**Comment by mtrang1263** [comment]
> What are the appropriate changes associated with this comment?
> 
> ```
> Will plan on adding appropriate changes after #549 is merged.
> ```
> 
> Since this PR is ready to go in first, should we just add the scaffolding as part of #549? Or are you envisioning a separate PR?


We'll be able to use the flags/plumbing that's being done in https://github.com/CMSgov/bcda-app/pull/549. I think it'll be pretty quick to rebase these changes on top so we wouldn't need another PR.

File: bcda/api/v2/api.go [hunk 4]
```diff
@@ -88,3 +95,114 @@ func BulkGroupRequest(w http.ResponseWriter, r *http.Request) {
...
+
+	Produces:
+	- application/fhir+json
+
+	Schemes: http, https
+
+	Responses:
+		200: MetadataResponse
+*/
+func Metadata(w http.ResponseWriter, r *http.Request) {
```
Comment (msnook) [inline bcda/api/v2/api.go:114]: Does the new endpoint need to be registered in the router?  For example: https://github.com/CMSgov/bcda-app/blob/master/bcda/web/router.go#L40-L45

**Comment by mtrang1263** [comment]
Verified that the two bdt-scan metadata checks are now passing:

```
 ● Metadata
   ● CapabilityStatement
     ✔ The CapabilityStatement instantiates the bulk-data CapabilityStatement
     ✔ Check if "patient-export" operation is defined in the CapabilityStatement
     ✔ Check if "group-export" operation is defined in the CapabilityStatement
```

**Comment by mtrang1263** [comment]
With the latest revision, we're passing all of the metadata checks:
```
 ● Metadata
   ● CapabilityStatement
     ✔ The CapabilityStatement instantiates the bulk-data CapabilityStatement
     ✔ Includes the token endpoint in the CapabilityStatement
     ✘ Check if "export" operation is defined in the CapabilityStatement
       └─ The "export" operation is not supported by this server
     ✔ Check if "patient-export" operation is defined in the CapabilityStatement
     ✔ Check if "group-export" operation is defined in the CapabilityStatement
   ● Well Known SMART Configuration
     ! Includes token_endpoint definition
       ├─ This test verifies that the server provides a /.well-known/smart-configuration and that a token_endpoint property is declared within that file.
       └⯈ No WellKnown JSON found at "http://host.docker.internal:3000/api/v2/.well-known/smart-configuration"

6 tests executed in 106 ms
     succeeded tests: 4
        failed tests: 0
     not implemented: 0
 not supported tests: 1
  generated warnings: 1
```

**Review by msnook — APPROVED** [review]
🎸

**Review by jeffcorpuz — APPROVED** [review]
Art

**Pull request merged**